fs-index/src/index.rs

```diff
@@ -1,1120 +1,436 @@
-) -> HashMap<CanonicalPathBuf, IndexEntry<Id>>
+/// Represents the index of resources in a directory.
+///
+/// [`ResourceIndex`] provides functionality for managing a directory index,
+/// including tracking changes, and querying resources.
+///
+/// #### Reactive API
+/// - [`ResourceIndex::update_all`]: Method to update the index by rescanning
+///   files and returning changes (additions/deletions/updates).
+///
+/// #### Snapshot API
+/// - [`ResourceIndex::get_resources_by_id`]: Query resources from the index by
+///   ID.
+/// - [`ResourceIndex::get_resource_by_path`]: Query a resource from the index
+///   by its path.
+///
+/// #### Track API
+/// Allows for fine-grained control over tracking changes in the index
+/// - [`ResourceIndex::track_addition`]: Track a newly added file (checks if the
+///   file exists in the file system).
+/// - [`ResourceIndex::track_removal`]: Track the deletion of a file (checks if
+///   the file was actually deleted).
+/// - [`ResourceIndex::track_modification`]: Track an update on a single file.
+///
+/// ## Examples
+/// ```no_run
+/// use std::path::Path;
+/// use fs_index::{ResourceIndex, load_or_build_index};
+/// use dev_hash::Crc32;
+///
+/// // Define the root path
+/// let root_path = Path::new("animals");
+///
+/// // Build the index
+/// let index: ResourceIndex<Crc32> = ResourceIndex::build(root_path).expect("Failed to build index");
+/// // Store the index
+/// index.store().expect("Failed to store index");
+///
+/// // Load the stored index
+/// let mut loaded_index: ResourceIndex<Crc32> = load_or_build_index(root_path, false).expect("Failed to load index");
+///
+/// // Update the index
+/// loaded_index.update_all().expect("Failed to update index");
+///
+/// // Get a resource by path
+/// let _resource = loaded_index
+///     .get_resource_by_path("cat.txt")
+///     .expect("Resource not found");
+/// ```
+#[derive(Clone, Debug)]
+pub struct ResourceIndex<Id>
 where
-    Id: ResourceId,
+    Id: Eq + Hash,
 {
-    entries
-        .into_iter()
-        .filter_map(|(path_buf, entry)| {
-            let metadata = entry.metadata().ok()?;
-
-            let path = path_buf.as_canonical_path();
-            let result = scan_entry(path, metadata);
-            match result {
-                Err(msg) => {
-                    log::error!(
-                        "Couldn't retrieve metadata for {}:\n{}",
-                        path.display(),
-                        msg
-                    );
-                    None
-                }
-                Ok(entry) => Some((path_buf, entry)),
-            }
-        })
-        .collect()
+    /// The root path of the index (canonicalized)
+    pub(crate) root: PathBuf,
+    /// A map from resource IDs to resources
+    ///
+    /// Multiple resources can have the same ID (e.g., due to hash collisions
+    /// or files with the same content)
+    pub(crate) id_to_resources: HashMap<Id, Vec<IndexedResource<Id>>>,
```

Comment:
We should remove ids from the right side of the mapping:
```
id_to_paths: HashMap<Id, HashSet<IndexEntry>>
```

If it's possible to also remove timestamps from the right side of the mapping, without breaking `update_all`, then it would be even better:
```
id_to_paths: HashMap<Id, HashSet<Path>>
```

Reply:
Updated `ResourceIndex` to include the following mappings:
```rust
#[derive(Clone, Debug)]
pub struct ResourceIndex<Id>
where
    Id: Eq + Hash,
{
    /// The root path of the index (canonicalized)
    pub(crate) root: PathBuf,
    /// A map from resource IDs to paths
    ///
    /// Multiple resources can have the same ID (e.g., due to hash collisions
    /// or files with the same content)
    pub(crate) id_to_paths: HashMap<Id, HashSet<PathBuf>>,
    /// A map from resource paths to resources
    pub(crate) path_to_resource: HashMap<PathBuf, IndexedResource<Id>>,
}
```

Updated the custom serialization and deserialization functions accordingly. Also made changes to `update_all()` to reflect these updates.

Reply:
I have a question regarding `update_all()`. We currently have two mappings:
- path -> resource (id, path, modified time)
- id -> path

We allow multiple resources to have the same ID (either due to the same content or hash collisions in non-cryptographic hashes).

Here is the return type of `update_all()`:
```rust
/// Represents the result of an update operation on the ResourceIndex
#[derive(PartialEq, Debug)]
pub struct IndexUpdate<Id: ResourceId> {
    /// Resources that were added during the update
    added: HashMap<Id, IndexedResource<Id>>,
    /// Resources that were removed during the update
    removed: HashSet<Id>,
}
```
If we have an index where three resources map to the same ID and one of these resources is removed, should we return the ID as if it was removed? It might be confusing since the ID still exists in the index.

Reply:
@tareknaser good question. Please, register new issue for this. In the opposite scenario, when new duplicate appeared, I would expect that we mention it in `added` field, but not in `removed` as we do with renamings. But if a duplicate is removed, we might need special field in the `IndexUpdate` structure.

Reply:
By the way, I think `added` field should be a mapping into collections of paths:
```
added: HashMap<Id, HashSet<PathBuf>>
```

On other hand, providing users with the timestamps could be also a good idea..

Reply:
> On other hand, providing users with the timestamps could be also a good idea..


I think we should include the modified time of added resources since this covers both newly created and modified resources.

For modified resources, users can compare the path of the updated resource with the resource of the same path from the previous state. The modified time can help users in some cases make informed decisions.

Reply:
@tareknaser this sounds like what our library should do, not users. But timestamps could be useful for sorting resources by date, so let's keep them for the moment.